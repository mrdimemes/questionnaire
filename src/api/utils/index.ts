export { default as getFetchError } from "./getFetchError";