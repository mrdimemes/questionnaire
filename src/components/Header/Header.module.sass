@use "src/sass/foundation/colors"
@use "src/sass/foundation/breakpoints"
@use "src/sass/foundation/mixins"

.body
  background-color: colors.$main-light
  font-size: 2em
  position: fixed
  top: 0
  left: 0
  right: 0
  height: 60px

.wrapper
  height: 100%
  display: flex
  justify-content: space-between
  align-items: center

.logo
  font-weight: 900
  text-transform: uppercase
  @include mixins.unselectable

.interactive
  display: flex
  align-items: center
  column-gap: 45px

.navigation
  display: flex
  align-items: center
  column-gap: 30px

.link
  display: block

.activeLink
  font-weight: 800

@media (max-width: breakpoints.$small-screen)
  .body
    font-size: 1.8em
    height: 50px
