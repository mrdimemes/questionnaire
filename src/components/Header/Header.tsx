import classNames from "classnames";

import { withTheme } from "src/HOCs";

import { Menu } from "./Menu";
import styles from "./Header.module.sass";

import type { HeaderProps } from "./types";


const Header = ({ className }: HeaderProps) => {
  return (
    <header className={classNames(styles.Header, className)}>
      <div className={classNames(styles.wrapper, "wrapper")}>
        <div className={styles.logo}>Questions!</div>
        <Menu />
      </div>
    </header>
  );
};

export default withTheme(Header, styles);