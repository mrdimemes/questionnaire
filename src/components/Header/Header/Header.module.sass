@use "src/sass/foundation/colors"
@use "src/sass/foundation/breakpoints"
@use "src/sass/foundation/mixins"

.body
  background-color: colors.$main-light
  font-size: 2em
  position: fixed
  top: 0
  left: 0
  right: 0
  height: 60px
  @media (max-width: breakpoints.$small-screen)
    font-size: 1.8em
    height: 50px
  @media (max-width: breakpoints.$tablet)
    font-size: 1.6em
    height: 40px

.wrapper
  display: flex
  justify-content: space-between
  align-items: center

.logo
  font-weight: 900
  text-transform: uppercase
  @include mixins.unselectable
