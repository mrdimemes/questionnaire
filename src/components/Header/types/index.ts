export type { HeaderProps } from "./HeaderProps";