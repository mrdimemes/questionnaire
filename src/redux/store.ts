import { combineReducers, configureStore } from "@reduxjs/toolkit";
import type { PreloadedState } from '@reduxjs/toolkit'
import { authSlice, tagsSlice, browseSlice, themeSlice } from "./slices";

const rootReducer = combineReducers({
  tags: tagsSlice,
  browse: browseSlice,
  auth: authSlice,
  theme: themeSlice
});

export const setupStore = (preloadedState?: PreloadedState<RootState>) => {
  return configureStore({
    reducer: rootReducer,
    preloadedState
  })
}

const store = setupStore();

export default store;
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
export type AppDispatch = AppStore['dispatch'];