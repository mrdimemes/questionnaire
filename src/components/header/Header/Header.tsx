import classNames from "classnames";
import { useThemeSelector } from "src/redux/hooks";
import { getThemeStyle } from "src/redux/slices/themeSlice";

import { Menu } from "../Menu";

import styles from "./Header.module.sass";

const Header = () => {
  const currentTheme = useThemeSelector();

  return (
    <header className={classNames(
      styles.body,
      getThemeStyle(styles, currentTheme),
    )}>
      <div className={`wrapper ${styles.wrapper}`}>
        <div className={styles.logo}>Questions!</div>
        <Menu />
      </div>
    </header>
  );
};

export default Header;