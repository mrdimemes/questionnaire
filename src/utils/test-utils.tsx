import React, { PropsWithChildren } from "react";
import { BrowserRouter } from "react-router-dom";
import { render, RenderOptions } from "@testing-library/react";
import { PreloadedState } from "@reduxjs/toolkit";
import { Provider } from "react-redux";
import { setupStore, AppStore, RootState } from "src/redux";

interface ExtendedRenderOptions extends Omit<RenderOptions, "queries"> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export function renderWithRouters(
  ui: React.ReactElement,
  { ...renderOptions }: ExtendedRenderOptions = {},
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <BrowserRouter>{children}</BrowserRouter>;
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

export function renderWithProvidersAndRouters(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {},
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>
      <BrowserRouter>{children}</BrowserRouter>
    </Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}