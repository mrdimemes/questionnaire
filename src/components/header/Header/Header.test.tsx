import { screen } from "@testing-library/react";
import { renderWithProvidersAndRouters } from "src/utils/test-utils";
import { setupStore } from "src/redux/store";
import Header from "./Header";

describe("Header component", () => {
  test("Menu renders without user in the store", () => {
    const store = setupStore({ auth: { user: null } })
    renderWithProvidersAndRouters(<Header />, { store });
    expect(screen.getByText(/questions!/i)).toBeInTheDocument();
    expect(screen.getByRole("navigation")).toBeInTheDocument();
    expect(screen.getByText(/главная/i)).toBeInTheDocument();
    expect(screen.getByText(/тэги/i)).toBeInTheDocument();
    expect(screen.getByText(/опросы/i)).toBeInTheDocument();
    screen.findAllByRole("button")
      .then((buttons) => expect(buttons).toHaveLength(3))
  });

  test("Menu renders with user in the store", () => {
    const store = setupStore({
      auth: {
        user: { name: "TestName", email: "test@email.com", isAdmin: false }
      }
    })
    renderWithProvidersAndRouters(<Header />, { store });
    expect(screen.getByText(/questions!/i)).toBeInTheDocument();
    expect(screen.getByText("TestName")).toBeInTheDocument();
    expect(screen.getByRole("navigation")).toBeInTheDocument();
    expect(screen.getByText(/главная/i)).toBeInTheDocument();
    expect(screen.getByText(/тэги/i)).toBeInTheDocument();
    expect(screen.getByText(/опросы/i)).toBeInTheDocument();
    screen.findAllByRole("button")
      .then((buttons) => expect(buttons).toHaveLength(4))
  });

  test("Header snapshot", () => {
    const { container } = renderWithProvidersAndRouters(<Header />);
    expect(container).toMatchSnapshot();
  });
});