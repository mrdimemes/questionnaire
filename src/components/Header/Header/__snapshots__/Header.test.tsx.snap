// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component Header snapshot 1`] = `
<div>
  <header
    class="body theme_light"
  >
    <div
      class="wrapper wrapper"
    >
      <div
        class="logo"
      >
        Questions!
      </div>
      <div
        class="body"
      >
        <div
          class="interactive theme_light"
        >
          <nav
            class="body"
          >
            <a
              class="link"
              href="/"
            >
              Главная
            </a>
            <a
              class="link"
              href="/questionnaires"
            >
              Опросы
            </a>
            <a
              class="link"
              href="/tags"
            >
              Тэги
            </a>
          </nav>
          <div
            class="meta"
          >
            <button
              class="body theme_light body"
            >
              <img
                alt="switch theme"
                class="icon"
                src="light-mode-icon.svg"
              />
            </button>
            <div
              class="body"
            >
              <button
                class="body theme_light"
              >
                Войти
              </button>
            </div>
          </div>
        </div>
        <button
          class="body theme_light body"
        >
          <div
            class="icon theme_light"
          />
        </button>
      </div>
    </div>
  </header>
</div>
`;
